service/waterfall.go

```diff
@@ -546,51 +668,63 @@ func (uis *UIServer) waterfallPage(w http.ResponseWriter, r *http.Request) {
 		skip = 0
 	}
 
-	finalData := waterfallData{}
+	variantQuery := strings.TrimSpace(r.URL.Query().Get(BVFilterParam))
 
 	// first, get all of the versions and variants we will need
-	vvData, err := getVersionsAndVariants(skip,
-		VersionItemsToCreate, project)
+	vvData, err := getVersionsAndVariants(
+		skip, VersionItemsToCreate, project, variantQuery,
+	)
+
+	if err != nil {
+		uis.LoggedError(w, r, http.StatusNotFound, err)
+		return
+	}
 
+	finalData, err := waterfallDataAdaptor(vvData, project, skip, variantQuery)
 	if err != nil {
 		uis.LoggedError(w, r, http.StatusInternalServerError, err)
 		return
 	}
-	var wfv waterfallVersions = vvData.Versions
 
-	sort.Sort(wfv)
-	finalData.Versions = wfv
+	uis.WriteHTML(w, http.StatusOK, struct {
+		Data     waterfallData
+		JiraHost string
+		ViewData
+	}{finalData, uis.Settings.Jira.Host, uis.GetCommonViewData(w, r, false, true)}, "base", "waterfall.html", "base_angular.html", "menu.html")
+}
 
-	sort.Sort(vvData.BuildVariants)
-	rows := []waterfallRow{}
-	for _, bv := range vvData.BuildVariants {
-		rows = append(rows, vvData.Rows[bv.Id])
+func (restapi restAPI) getWaterfallData(w http.ResponseWriter, r *http.Request) {
+	projCtx := MustHaveRESTContext(r)
+	project, err := projCtx.GetProject()
+	if err != nil || project == nil {
+		restapi.WriteJSON(w, http.StatusNotFound, responseError{Message: "error finding project"})
+		return
 	}
-	finalData.Rows = rows
 
-	// compute the total number of versions that exist
-	finalData.TotalVersions, err = version.Count(version.ByProjectId(project.Identifier))
+	query := r.URL.Query()
+	// 0 is fail safe defailt value
+	skip, _ := strconv.Atoi(query.Get(SkipQueryParam))
```

Comment:
You should check the second value here to see if the query string is a valid int

Reply:
When error happens, the value will be initialized with zero-value for given type. In my case it will be 0, which is what I want.

Reply:
It will, but I think it's more appropriate to error if the client sends something non-numeric here than to use 0